### PR TITLE
Docs: Show primary story description and headline in autodocs

### PR DESCRIPTION
**Pull request by tmeasday**
Issue: N/A

Telescoping on https://github.com/storybookjs/storybook/pull/20559

## What I did

Added a render option `forceInitialArgs` which means a story is rendered with it's initial args no matter what and doesn't respond to arg updates. Use this in the `<Stories/>` block, and switch to showing the primary story there.

The idea is that now in Autodocs we show:

- Primary story, reactive
- Controls (to control ☝️)
- All stories (inc. primary) with descriptions, unreactive.

image

## How to test

- Look at tests + stories
- Try autodocs!

### CODE REVIEW

File: code/lib/preview-api/src/modules/preview-web/PreviewWeb.test.ts [hunk 1]
```diff
@@ -1131,6 +1131,39 @@ describe('PreviewWeb', () => {
             'story-element'
           );
         });
+
+        it('does not re-renders the story when ignoreArgsUpdates=true', async () => {
```
Comment (JReinhold) [inline code/lib/preview-api/src/modules/preview-web/PreviewWeb.test.ts:1135]:
```suggestion
        it('does not re-render the story when ignoreArgsUpdates=true', async () => {
```

File: code/ui/blocks/src/components/Story.stories.tsx [hunk 2]
```diff
@@ -48,9 +48,38 @@ export const IFrame = {
...
 
+export const IgnoreArgsUpdates = {
+  args: {
+    storyExport: ButtonStories.Primary,
+    inline: true,
+    autoplay: true,
+    ignoreArgsUpdates: true,
+    renderStoryToElement,
+  },
+  play: async ({ args, canvasElement, loaded }: PlayFunctionContext<WebRenderer>) => {
```
Comment (JReinhold) [inline code/ui/blocks/src/components/Story.stories.tsx:59]:
```suggestion
  // test that it ignores updated args by emitting an arg update and assert that it isn't reflected in the DOM
  play: async ({ args, canvasElement, loaded }: PlayFunctionContext<WebRenderer>) => {
```

**Review by JReinhold — APPROVED** [review]
Looks good! 👍

**Comment by tmeasday** [comment]
Ok, now this is ready for re-review. Sorry for the early review @JReinhold. I changed the name and added a test for the "initial" behaviour.

I wasn't sure if we wanted to add stories for the `Stories` and `DocsStory` components. I sort of feel like they'll be duplicative and it'll be hard to actually test it's working outside of more copies of the same play function.

**Comment by tmeasday** [comment]
@JReinhold so the e2e test is failing due to this query:

https://github.com/storybookjs/storybook/blob/c22aac8e5bb8a85bb81fd58db1d9664a9662e6c3/code/e2e-tests/addon-docs.spec.ts#L49

I guess that's because the `Story` component generates a DOM id based on the story's id: 

https://github.com/storybookjs/storybook/blob/c22aac8e5bb8a85bb81fd58db1d9664a9662e6c3/code/ui/blocks/src/components/Story.tsx#L60

What shall we do about this do you think?

**Comment by shilman** [comment]
I think the primary story should have a special ID. not sure what though. Not sure what though `<id>--primary`?--or how it would impact anything

**Comment by tmeasday** [comment]
Ok, that's make sense. I did that.

**Comment by tmeasday** [comment]
@valentinpalkovic the failure on this branch comes from the angular renderer which doesn't appear to be able to render the same story more than once on the screen:

image

Is that surprising to you? If it's expected would it be easy to change/fix?

**Comment by valentinpalkovic** [comment]
Fixed in https://github.com/storybookjs/storybook/pull/20559 :)

**Comment by tmeasday** [comment]
Oh amazing!

**Comment by JReinhold** [comment]
@valentinpalkovic 

> Fixed in https://github.com/storybookjs/storybook/pull/20559 :)

@tmeasday

Are you sure about that Valentin? The issue you and I discussed a few days ago, was about Angular being unable to render any multiple stories twice, like it does in docs. But this here is specifically about rendering _the same_ story twice.

Maybe you've already taken care of this since our last conversation, then please ignore me.

File: code/ui/blocks/src/blocks/Story.stories.tsx [hunk 3]
```diff
@@ -199,3 +200,12 @@ export const WithInteractionsAutoplayInStory: Story = {
     chromatic: { delay: 500 },
   },
 };
+
+export const IgnoreArgsUpdates: Story = {
```
Comment (JReinhold) [inline code/ui/blocks/src/blocks/Story.stories.tsx:204]:
```suggestion
export const ForceInitialArgs: Story = {
```

**Review by JReinhold — APPROVED** [review]
LGTM! 👍

**Comment by valentinpalkovic** [comment]
> @valentinpalkovic
> 
> > Fixed in #20559 :)
> 
> @tmeasday
> 
> Are you sure about that Valentin? The issue you and I discussed a few days ago, was about Angular being unable to render any multiple stories twice, like it does in docs. But this here is specifically about rendering _the same_ story twice.
> 
> Maybe you've already taken care of this since our last conversation, then please ignore me.

Angular components are rendered as Standalone components in https://github.com/storybookjs/storybook/pulls. Therefore, the same component can be rendered multiple times without any issues.

**Comment by tmeasday** [comment]
I changed the base of this https://github.com/storybookjs/storybook/pull/20559, let's find out for sure!

**Comment by valentinpalkovic** [comment]
@tmeasday Hmm. Still, it does not work. let me take a quick look. Shouldn't be that hard to fix.

**Comment by chakAs3** [comment]
In my opinion, The Main Story, Default or Playground should be generated as the Base Story from the component definition only ( with default values ) and be independent of all Stories. 
Base/Default Story is the Interactive Playground that can match the state of all the stories and not only the first one.

I used the same technique in most Storybook projects 

image

so the Playground on where you mix and match different states of your component, and mostly the following stories are complete illustrations of Variants so the user can see a max of the component's states. 

i have been raising some issues since the update.

**Comment by JReinhold** [comment]
> In my opinion, The Main Story, Default or Playground should be generated as the Base Story from the component definition only ( with default values ) and be independent of all Stories. Base/Default Story is the Interactive Playground that can match the state of all the stories and not only the first one.
> 
> I used the same technique in most Storybook projects
> 
> image
> 
> so the Playground on where you mix and match different states of your component, and mostly the following stories are complete illustrations of Variants so the user can see a max of the component's states.
> 
> i have been raising some issues since the update.

I don't disagree with you, but in the current data model I don't think that's possible. There's no such thing as displaying a story only based on the meta information, only actual stories can be displayed.

And what would you expect to be shown with the following stories file?
```tsx
export default {
  title: 'My Button'
}

export const Primary = {
  render: () => <button>hello</button>
}
```

Without a `component` or a `render` function on meta, we can't render anything solely based on the meta information.

**Comment by chakAs3** [comment]
> > In my opinion, The Main Story, Default or Playground should be generated as the Base Story from the component definition only ( with default values ) and be independent of all Stories. Base/Default Story is the Interactive Playground that can match the state of all the stories and not only the first one.
> > I used the same technique in most Storybook projects
> > image
> > so the Playground on where you mix and match different states of your component, and mostly the following stories are complete illustrations of Variants so the user can see a max of the component's states.
> > i have been raising some issues since the update.
> 
> I don't disagree with you, but in the current data model I don't think that's possible. There's no such thing as displaying a story only based on the meta information, only actual stories can be displayed.
> 
> And what would you expect to be shown with the following stories file?
> 
> ```tsx
> export default {
>   title: 'My Button'
> }
> 
> export const Primary = {
>   render: () => <button>hello</button>
> }
> ```
> 
> Without a `component` or a `render` function on meta, we can't render anything solely based on the meta information.

I know that's why i said generated Story we generate what we need to render the Story  based on the Component + meta info 
I need a component and args to render and if there is a Story context that will be great more info, 

from your code, I expect a  hello button

**Comment by GuillaumeNury** [comment]
@tmeasday hello! What is the motivation of changing the default value of `includePrimary` ?
I cannot find a way to change this value.

**Comment by JReinhold** [comment]
> @tmeasday hello! What is the motivation of changing the default value of `includePrimary` ? I cannot find a way to change this value.

1. It seemed less magical and more obvious to us that the list of Stories shows all stories, and didn't "hide" anything.
2. Our new support for story-level descriptions (along side component-level descriptions) needed a place to show the description for the primary story, and showing it in the list with the rest of the stories made this a non-issue.

I believe you can get back the old behavior by providing your own custom `DocsPage` that largely copies the existing one:

https://storybook.js.org/docs/7.0/react/writing-docs/docs-page#write-a-custom-template

```ts
// .storybook/preview.js

import React from 'react';
import { Title } from './Title';
import { Subtitle } from './Subtitle';
import { Description } from './Description';
import { Primary } from './Primary';
import { Controls } from './Controls';
import { Stories } from './Stories';

export const parameters = {
  ...
  docs: {
    page: () => (
	  <>
	    <Title />
	    <Subtitle />
	    <Description />
	    <Primary />
	    <Controls />
	    <Stories includePrimary={false} />
	  </>
	);
  },
};
```

**Comment by GuillaumeNury** [comment]
Ok. We use Angular, I can see that I should dig around `mdx` file ([doc here](https://storybook.js.org/docs/7.0/angular/writing-docs/docs-page#write-a-custom-template)).
Our design system is documented using [ZeroHeight](https://zeroheight.com/). We display only one story with its controls on each page ([exemple here](https://prisme.lucca.io/94310e217/p/07849c-radio) which exposes [this story](https://lucca-front.lucca.io/storybook/?path=/docs/documentation-forms-radios-basic--docs)). Currently, we hide additional stories using CSS 🙈

**Comment by GuillaumeNury** [comment]
@JReinhold I played with custom Docs template and it worked like a charm. Thank you!

**Pull request merged**